src/uwtools/config.py

```diff
@@ -331,14 +331,19 @@ def update_values(self, new_dict, dict_to_update=None):
         '''
 
         if dict_to_update is None:
-            dict_to_update = self.data
+            dict_to_update = self.data or {}
```

Comment:
Apparently `self.data` is `None` before we initialize it. 

I'll be opening an Issue about the need for any testing at all of this method.

Reply:
Can you share what error you encountered that led to this change? I can't see how `self.data` could be `None`. `UserDict` initializes it to `{}`, so something must be changing it to `None`, and that seems like violation of the invariant that `self.data` should be a `dict`, per the `UserData` documentation:
> The instance’s contents are kept in a regular dictionary, which is accessible via the `data` attribute of `UserDict` instances.
If you get the stacktrace of the failure, it might tell us where the invalid `None` value is originating.

I'll take this opportunity to sound the first note of a drum I'll be beating in the future: We should incorporate `mypy` typing into this codebase. In addition to providing insanely useful documentation to anyone reading the code, it also catches cases where unexpected `None` values arise and force you to deal with them. I can't overstate how valuable I've found it both in catching and avoiding bugs, but also in making it clear when my code is poorly designed: If it's hard to write a correct (and useful!) type for a function, argument, variable, etc., it's probably too complicated. I'll be very happy to help do the work!

@fgabelmannjr I heard you mention that you were more steeped in the JS ecosystem than Python, so I wonder if you've also used TypeScript and what you think of it.

Reply:
@maddenp-noaa Good call all around here to sound drums. I'm looking back at the offender of this, and it's definitely a caller breach of contract...check out what dumb thing I did in the UW integration.

```
    # Load pre-defined settings for the physics suite
    suite_config = uw_config.YAMLConfig(
        config_path=FV3_NML_YAML_CONFIG_FP,
        )
    # Update the config object to include only the chosen suite data
    suite_config.data = suite_config.data.get(CCPP_PHYS_SUITE)

    # Combine the information from the settings dict and the default
    # suite, where items defined in settings take priority.
    suite_config.update_values(settings)
```

Now, the question is... given that this dumb move is totally allowed, how do we protect against it? I mean...should I have really done it this way? No, I had some icky feelings that it was bad when I did it. I mean, first pass it that I change the SRW integration piece to make sure it am setting the object to an actual dict with:

```
suite_config.data = suite_config.data.get(CCPP_PHYS_SUITE, {})
```

Is this the place where a setter comes into play? If so, would you mind guiding that process a bit more? I've never actually (knew that I) needed one before. Also, I'm not sure what that looks like given that `data` is a built-in public, default method and I want to be careful about overwriting native features when setting `data`.

Reply:
@christinaholtNOAA Probably disappointingly, my answer has to be "It depends". I mean, there are lots of interesting things we could do to make things safer and enforce desirable constraints/invariants, but today I lack the background to know what those constraints/invariants would be. As a first guess, we might not want to inherit from `UserDict` but, instead, design a class with a _subset_ of `dict` semantics but that enforces behavior we want to rely on. Rather than having a public `data` member, it might manage a private `dict` and force callers to modify it via method calls (including e.g. `__setattr__()` to support `dict` `a[b] = c` setter semantics), which would never allow wholesale replacement of the private `dict` like we saw here -- the private `dict` would _always_ be a `dict`.

But I'd probably benefit from a 15-minute discussion about what that class should be/do, looking ahead and making sure nobody sees any obvious deficiencies before we make it. And I don't think it would be hard at all, by the way -- probably just tweaks to what we've already go. But it'd be worth trying to poke holes in the idea.